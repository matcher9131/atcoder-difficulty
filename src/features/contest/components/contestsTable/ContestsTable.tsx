import type { ReactElement } from "react";
import clsx from "clsx";

export type ContestsTableProps = {
    readonly contestsTableHeader: ReactElement;
    readonly gridColsClassName: string;
    readonly contestRows: readonly ReactElement[];
};

export const ContestsTable = ({
    contestsTableHeader,
    gridColsClassName,
    contestRows,
}: ContestsTableProps): ReactElement => {
    return (
        <table className={clsx("w-full", "grid", gridColsClassName)}>
            {contestsTableHeader}
            <tbody className="contents">{contestRows}</tbody>
        </table>
    );
};
