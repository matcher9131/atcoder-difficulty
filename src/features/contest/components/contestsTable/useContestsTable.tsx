import { useAtom } from "jotai";
import type { ContestType } from "../../types/contestType";
import { contestIdsByTypeAtom } from "../../dict/contests";
import type { ContestsTableProps } from "./ContestsTable";
import { ContestRowContainer } from "../contestRow";
import { getGridColsClassName, getNumProblems } from "../../functions/numProblems";
import { paginationValueAtom } from "../../../pagination/model/paginations";
import { ContestsTableHeader } from "../contestsTableHeader";
import { itemsPerPageAtom } from "../../../pagination/model/itemsPerPage";

export const useContestsTable = (contestType: ContestType): ContestsTableProps => {
    const numProblems = getNumProblems(contestType);
    const contestsTableHeader = <ContestsTableHeader numProblems={numProblems} />;
    const gridColsClassName = getGridColsClassName(contestType);
    const [contestIds] = useAtom(contestIdsByTypeAtom(contestType));
    const [pageIndex] = useAtom(paginationValueAtom(contestType));
    const [itemsPerPage] = useAtom(itemsPerPageAtom(contestType));
    const contestRows = contestIds
        .slice(itemsPerPage * pageIndex, itemsPerPage * (pageIndex + 1))
        .map((contestId) => <ContestRowContainer key={contestId} contestId={contestId} numProblems={numProblems} />);
    return { contestsTableHeader, gridColsClassName, contestRows };
};
