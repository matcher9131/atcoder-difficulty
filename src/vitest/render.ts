import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
