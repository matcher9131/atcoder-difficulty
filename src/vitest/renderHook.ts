import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
