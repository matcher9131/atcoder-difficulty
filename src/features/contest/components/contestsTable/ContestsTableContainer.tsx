import type { ReactNode } from "react";
import type { ContestType } from "../../types/contestType";
import { useContestsTable } from "./useContestsTable";
import { ContestsTable } from "./ContestsTable";

type ContestsTableContainerProps = {
    readonly contestType: ContestType;
};

export const ContestsTableContainer = ({
    contestType,
}: ContestsTableContainerProps): ReactNode => {
    const props = useContestsTable(contestType);
    return <ContestsTable {...props} />;
};
