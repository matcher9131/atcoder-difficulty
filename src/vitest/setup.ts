import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
