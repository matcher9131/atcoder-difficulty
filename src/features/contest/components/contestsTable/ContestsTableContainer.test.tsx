import { describe, it, expect, vi, beforeEach } from "vitest";
import { screen } from "@testing-library/react";
import { render } from "../../../../vitest/render";
import { ContestsTableContainer } from "./ContestsTableContainer";
import { useContestsTable } from "./useContestsTable";
import { ContestsTable } from "./ContestsTable";

// src/features/contest/components/contestsTable/ContestsTableContainer.test.tsx

// Mock useContestsTable
vi.mock("./useContestsTable", () => ({
    useContestsTable: vi.fn(),
}));

// Mock ContestsTable
vi.mock("./ContestsTable", () => ({
    ContestsTable: vi.fn(() => <div data-testid="contests-table-mock" />),
}));

describe("ContestsTableContainer", () => {
    const dummyContestType = "abc";

    const mockProps = {
        contestsTableHeader: (
            <thead>
                <tr>
                    <th>Header</th>
                </tr>
            </thead>
        ),
        gridColsClassName: "grid-cols-2",
        contestRows: [
            <tr key="1">
                <td>Row 1</td>
            </tr>,
        ],
    };

    beforeEach(() => {
        vi.clearAllMocks();
    });

    it("calls useContestsTable with contestType", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestsTable as any).mockReturnValue(mockProps);
        render(<ContestsTableContainer contestType={dummyContestType} />);
        expect(useContestsTable).toHaveBeenCalledWith(dummyContestType);
    });

    it("renders ContestsTable with props from useContestsTable", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestsTable as any).mockReturnValue(mockProps);
        render(<ContestsTableContainer contestType={dummyContestType} />);
        expect(screen.getByTestId("contests-table-mock")).toBeInTheDocument();
        expect(ContestsTable).toHaveBeenCalledWith(mockProps, undefined);
    });

    it("matches snapshot", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestsTable as any).mockReturnValue(mockProps);
        const { container } = render(<ContestsTableContainer contestType={dummyContestType} />);
        expect(container).toMatchSnapshot();
    });
});
