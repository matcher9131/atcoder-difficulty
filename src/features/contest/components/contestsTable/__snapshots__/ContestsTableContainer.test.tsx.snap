// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContestsTableContainer > matches snapshot 1`] = `
<div>
  <div
    data-testid="contests-table-mock"
  />
</div>
`;
